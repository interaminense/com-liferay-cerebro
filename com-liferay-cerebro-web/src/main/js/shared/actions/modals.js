export const actionTypes = {
	CLOSE_ALL_MODALS: 'CLOSE_ALL_MODALS',
	CLOSE_MODAL: 'CLOSE_MODAL',
	OPEN_MODAL: 'OPEN_MODAL'
};

export const modalTypes = {
	CREATE_MAPPING_MODAL: 'CREATE_MAPPING_MODAL',
	CSV_PREVIEW_MODAL: 'CSV_PREVIEW_MODAL',
	EDIT_CARD_MODAL: 'EDIT_CARD_MODAL',
	LOADING_MODAL: 'LOADING_MODAL',
	MATCH_FIELDS_MODAL: 'MATCH_FIELDS_MODAL',
	SELECT_ITEMS_MODAL: 'SELECT_ITEMS_MODAL',
	TEST: 'TEST'
};

export function close() {
	return {
		type: actionTypes.CLOSE_MODAL
	};
}

export function closeAll() {
	return {
		type: actionTypes.CLOSE_ALL_MODALS
	};
}

export function open(type, props = {}, options = {}) {
	const {closeOnBlur = true} = options;

	return {
		payload: {
			closeOnBlur,
			props,
			type
		},
		type: actionTypes.OPEN_MODAL
	};
}