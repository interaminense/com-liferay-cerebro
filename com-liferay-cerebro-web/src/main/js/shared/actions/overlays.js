export const actionTypes = {
	HIDE_OVERLAY: 'HIDE_OVERLAY',
	SHOW_OVERLAY: 'SHOW_OVERLAY'
};

export function hide(id) {
	return {
		payload: {
			id
		},
		type: actionTypes.HIDE_OVERLAY
	};
}

export function show(options) {
	const {
		alignment,
		containerClass,
		context,
		forceAlignment,
		id,
		offset,
		onMouseEnter,
		onMouseLeave,
		onOutsideClick,
		parentId
	} = options;

	return {
		payload: {
			alignment,
			containerClass,
			context,
			forceAlignment,
			id,
			offset,
			onMouseEnter,
			onMouseLeave,
			onOutsideClick,
			parentId
		},
		type: actionTypes.SHOW_OVERLAY
	};
}