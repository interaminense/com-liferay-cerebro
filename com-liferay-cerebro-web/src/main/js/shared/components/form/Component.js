import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';

import {checkRequired} from '../../util/form-validators';

class FormComponent extends Component {
	attached() {
		this.on('valueChanged', this.informChange);

		const {ancestorForm} = this.context;

		if (ancestorForm) {
			ancestorForm.addFormElement(this);
		}
	}

	detached() {
		const {ancestorForm} = this.context;

		if (ancestorForm) {
			ancestorForm.removeFormElement(this);
		}
	}

	getFormElementName() {
		return this.props.name;
	}

	getFormElementValue() {
		const {name, processValue} = this.props;

		let {value} = this.state;

		if (processValue) {
			value = processValue(value, name);
		}

		return value;
	}

	@autobind
	informChange() {
		if (this.context.ancestorForm) {
			this.context.ancestorForm.informChange(this.props.name);
		}
	}

	isRequired() {
		const {validator} = this.props;

		return validator && checkRequired(validator);
	}

	validateFormElement() {
		return true;
	}
}

FormComponent.PROPS = {
	name: Config.string().required(),
	processValue: Config.func()
};

FormComponent.STATE = {
	value: Config.any()
};

export {Config} from 'metal-jsx';

export default FormComponent;