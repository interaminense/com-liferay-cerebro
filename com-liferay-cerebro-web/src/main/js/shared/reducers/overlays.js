import {createReducer} from 'redux-toolbox';
import {OrderedMap, Map} from 'immutable';

import {actionTypes} from '../actions/overlays';

export default createReducer(OrderedMap(), {
	[actionTypes.HIDE_OVERLAY](state, action) {
		return state.delete(action.payload.id);
	},

	[actionTypes.SHOW_OVERLAY](state, action) {
		const {
			alignment,
			containerClass,
			context,
			forceAlignment,
			id,
			offset,
			onMouseEnter,
			onMouseLeave,
			onOutsideClick,
			parentId
		} = action.payload;

		return state.set(
			id,
			Map({
				alignment,
				containerClass,
				context,
				forceAlignment,
				id,
				offset,
				onMouseEnter,
				onMouseLeave,
				onOutsideClick,
				parentId
			})
		);
	}
});