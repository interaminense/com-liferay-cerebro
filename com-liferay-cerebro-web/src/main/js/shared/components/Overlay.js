import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';
import dom from 'metal-dom';
import {Align} from 'metal-position';
import {connect} from 'metal-redux';
import {debounce, get, isNil, noop, uniqueId} from 'lodash';

import overlayRegistry from 'shared/util/overlay-registry';
import {show, hide} from 'shared/actions/overlays';

export const ALIGNMENTS = [
	'topCenter',
	'topRight',
	'rightCenter',
	'bottomRight',
	'bottomCenter',
	'bottomLeft',
	'leftCenter',
	'topLeft'
];

const ALIGNMENTS_MAP = {
	bottomCenter: Align.BottomCenter,
	bottomLeft: Align.BottomLeft,
	bottomRight: Align.BottomRight,
	leftCenter: Align.LeftCenter,
	rightCenter: Align.RightCenter,
	topCenter: Align.TopCenter,
	topLeft: Align.TopLeft,
	topRight: Align.TopRight
};

class Overlay extends Component {
	created() {
		this._id = uniqueId();

		const {hideDelay, showDelay} = this.props;

		this.hide = debounce(() => (this.state.active = false), hideDelay);
		this.show = debounce(() => (this.state.active = true), showDelay);
	}

	detached() {
		if (this._bodyHandler) {
			this._bodyHandler.removeListener();
		}
	}

	rendered() {
		const active = this.getActive();

		if (!active && this._overlayVisible) {
			this._overlayVisible = false;

			this.removeContent();
		} else if (active) {
			this._overlayVisible = true;

			this.renderContent();
		}
	}

	disposed() {
		this.removeContent();
	}

	getActive() {
		const {active} = this.props;

		return !isNil(active) ? active : this.state.active;
	}

	@autobind
	handleMouseEnter() {
		const {overlay} = this.context;

		if (overlay) {
			overlay.onMouseEnter();
		}

		this.hide.cancel();

		this.show();
	}

	@autobind
	handleMouseLeave() {
		const {overlay} = this.context;

		if (overlay) {
			overlay.onMouseLeave();
		}

		this.show.cancel();

		this.hide();
	}

	@autobind
	handleOutsideClick(event) {
		if (!dom.contains(this.element, event.target)) {
			this.props.onOutsideClick();
		}
	}

	removeContent() {
		overlayRegistry.remove(this._id);

		this.props.hide(this._id);
	}

	renderContent() {
		const {
			alignment,
			containerClass,
			forceAlignment,
			offset,
			passedChildren,
			show
		} = this.props;

		overlayRegistry.add(this._id, passedChildren[1]);

		show({
			alignment: ALIGNMENTS_MAP[alignment],
			containerClass,
			context: this.context,
			forceAlignment,
			id: this._id,
			offset,
			onMouseEnter: this.handleMouseEnter,
			onMouseLeave: this.handleMouseLeave,
			onOutsideClick: this.handleOutsideClick,
			parentId: get(this, 'context.overlay.parentId')
		});
	}

	render() {
		const {passedChildren} = this.props;

		const trigger = passedChildren[0];

		trigger.props = {
			...trigger.props,
			active: this.getActive(),
			id: `overlay${this._id}`,
			onMouseEnter: this.handleMouseEnter,
			onMouseLeave: this.handleMouseLeave
		};

		return trigger;
	}
}

const STORE = {
	hide: Config.func().required(),
	show: Config.func().required()
};

Overlay.PROPS = {
	...STORE,
	active: Config.bool(),
	alignment: Config.oneOf(ALIGNMENTS).value('bottomCenter'),
	containerClass: Config.string(),
	forceAlignment: Config.bool().value(false),
	hideDelay: Config.number().value(400),
	offset: Config.number().value(8),
	onOutsideClick: Config.func().value(noop),
	passedChildren: Config.array(),
	showDelay: Config.number().value(400)
};

Overlay.STATE = {
	active: Config.bool().value(false)
};

export {Align} from 'metal-position';

export default connect(
	(state, {children}) => ({
		passedChildren: children
	}),
	{
		hide,
		show
	}
)(Overlay);