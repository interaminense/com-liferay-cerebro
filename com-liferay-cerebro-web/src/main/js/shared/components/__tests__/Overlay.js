jest.mock('shared/actions/overlays');
jest.mock('lodash');

import Component, {Config} from 'metal-jsx';

import * as actions from 'shared/actions/overlays';
import mockStore, {renderWithStore} from 'test/mock-store';
import Overlay from '../Overlay';

describe('Overlay', () => {
	let component;

	actions.hide.mockReturnValue({type: 'NOOP'});
	actions.show.mockReturnValue({type: 'NOOP'});

	afterEach(() => {
		if (component) {
			component.dispose();
		}

		jest.useFakeTimers();
	});

	it('should render the trigger', () => {
		component = new Overlay({
			children: [
				<button key="foo">{'trigger'}</button>,
				<div key="bar">{'content'}</div>
			],
			store: mockStore()
		});

		expect(component).toMatchSnapshot();
	});

	it.skip('should call show on mouse enter', done => {
		jest.useRealTimers();

		component = new Overlay({
			children: [
				<button key="foo">{'trigger'}</button>,
				<div key="bar">{'content'}</div>
			],
			showDelay: 5,
			store: mockStore()
		});

		expect(actions.show).not.toHaveBeenCalled();

		component.refs.child.handleMouseEnter();

		setTimeout(() => {
			expect(actions.show).toHaveBeenCalled();

			done();
		}, 30);
	});

	it.skip('should call hide on mouse leave', done => {
		jest.useRealTimers();

		component = new Overlay({
			children: [
				<button key="foo">{'trigger'}</button>,
				<div key="bar">{'content'}</div>
			],
			showDelay: 5,
			store: mockStore()
		});

		actions.hide.mockClear();

		expect(actions.hide).not.toHaveBeenCalled();

		component.refs.child.handleMouseLeave();

		setTimeout(() => {
			expect(actions.hide).toHaveBeenCalled();

			done();
		}, 30);
	});

	it('should call render content if the overlay content re-renders', () => {
		class TestComponent extends Component {
			render() {
				return (
					<Overlay active>
						<button>{'trigger'}</button>
						<div class="content">{this.props.content}</div>
					</Overlay>
				);
			}
		}

		TestComponent.PROPS = {
			content: Config.string().required()
		};

		component = renderWithStore(TestComponent, {
			content: 'foo'
		});

		expect(actions.show).toHaveBeenCalledTimes(1);

		component.props.content = 'bar';

		jest.runAllTimers();

		expect(actions.show).toHaveBeenCalledTimes(2);
	});
});