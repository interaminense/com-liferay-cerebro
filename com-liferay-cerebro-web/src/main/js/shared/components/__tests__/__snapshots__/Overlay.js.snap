// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay should render the trigger 1`] = `<button>trigger</button>`;
